📖  doc: Fix multigroup migration guide for v4 (#3403)
kubernetes-sigs/kubebuilder: docs/book/src/migration/multi-group.md

<h1>Note</h1>

While Kubebuilder will not scaffold out a project structure compatible
with multiple API groups in the same repository by default, it's possible

> to modify the default project structure to support it.

Reviewer: ```suggestion
to modify the default project structure to support it.

Note that the process mainly is to ensure that your  API(s) and controller(s) will be moved under new directories with their respective group name. 
```

Reviewer: What about we just clarifies as above mainly what we does in the process?

</aside>

Let's migrate the [CronJob example][cronjob-tutorial].

<aside class="note warning">
<h1>Instructions vary per project layout</h1>

You can verify the version by looking at the PROJECT file. The currently default and
recommended version is go/v4.

The layout go/v3 is **deprecated**, if you are using go/v3 it is recommended that you migrate to

> go/v4, however this documentation is still valid. [(More Info)][migration-guide].

Reviewer: Just a nit it is missing add the alias at the bottom, see in the preview: https://deploy-preview-3403--kubebuilder.netlify.app/migration/multi-group.html
![Screenshot 2023-05-16 at 22 04 57](https://github.com/kubernetes-sigs/kubebuilder/assets/7708031/3fc1ad3e-dac3-4cdc-b735-d4679e6561f6)

> This document is incompatible with go/v2. The documentation for go/v2 is still available in
> the [legacy documentation](https://book-v2.book.kubebuilder.io/migration/multi-group.html).
> 
> The layout go/v1 is incompatible with multi-group.

Reviewer: ```suggestion
```

It is very unlike someone been using v1/v2 if so they should check the old docs and not this one,
Anyway, if that does not clarify does we can address via chart or issues.

However, here I think we could try to make it more compreensive and help 90% of the audience.

> 

Reviewer: Could we ensure that we explain all with go/v4, which is the current default scaffold and then, at the bottom, we create a new section for the go/v3 layout? Also, could we add a WARNING note saying the go/v3 is deprecated?

Let's migrate the [CronJob example][cronjob-tutorial].

<aside class="note warning">
<h1>Using go/v3 or go/v4</h1>

The instructions vary depending on the project layout. You can verify the version by looking at
the PROJECT file.

> The layout go/v3 is deprecated, if you are usign go/v3 it is recommended that you migrate to go/v4.

Reviewer: ```suggestion
The layout go/v3 is **deprecated**, if you are using go/v3 it is recommended that you migrate to go/v4. [(More Info)][migration-guide]
```
Can we add the alias for the migration-guide at the bottom.

Also, note that the controllers will be created under `internal/controller/<group>` instead of `internal/controller`.

That is the reason why we moved the previously generated APIs to their respective locations in the new structure.
Remember to update the references in imports accordingly.

For envtest to install CRDs correctly into the test environment, the relative path to the CRD directory needs to be updated accordingly in each `internal/controller/<group>/suite_test.go` file. We need to add additional `".."` to our CRD directory relative path as shown below.

```go
    By("bootstrapping test environment")
    testEnv = &envtest.Environment{
        CRDDirectoryPaths: []string{filepath.Join("..", "..", "config", "crd", "bases")},
    }
```

The [CronJob tutorial][cronjob-tutorial] explains each of these changes in
more detail (in the context of how they're generated by Kubebuilder for
single-group projects).

[multi-group-issue]: https://github.com/kubernetes-sigs/kubebuilder/issues/923 "Kubebuilder Issue #923"
[cronjob-tutorial]: /cronjob-tutorial/cronjob-tutorial.md "Tutorial: Building CronJob"

> <h1>Using Go/v3</h1>

Reviewer: ```suggestion
### Using Go/v3
```

package v1
```

Then, we'll rename move our existing APIs into a new subdirectory, "batch":

```bash

> mv api/* api/batch

Reviewer: All steps still required rifght?
We need to create the dir with the group name
mv all api files to each and then remove api/

Author: We only need the mkdir, which I added back. The `mv api/* api/batch` will throw a warning that looks like `mv: cannot move 'batch' to a subdirectory of itself, 'batch/batch'`
 but that can be ignored safely.

```

After moving the APIs to a new directory, the same needs to be applied to the controllers. For go/v4:

```bash
mkdir internal/controller/batch
mv internal/controller/* internal/controller/batch/
```

> For go/v3:
> ```bash
> mkdir controller/batch
> mv controller/* controller/batch/
> ```

Reviewer: ````suggestion

<aside class="note">
<h1>If you are using the deprecated layout go/v3</h1>
Then, your layout has not the internal directory. So, you will move the controller(s) under a directory with the name of the API group which it is responsible for manage.
```bash
mkdir controller/batch
mv controller/* controller/batch/
```
````

[multi-group-issue]: https://github.com/kubernetes-sigs/kubebuilder/issues/923 "Kubebuilder Issue #923"
[cronjob-tutorial]: /cronjob-tutorial/cronjob-tutorial.md "Tutorial: Building CronJob"

<h1>Using Go/v4</h1>

Generally, we use the prefix for the API group as the directory name. We
can check `api/v1/groupversion_info.go` to find that out:

```go
// +groupName=batch.tutorial.kubebuilder.io
package v1
```

Then, we'll rename move our existing APIs into a new subdirectory, "batch":

```bash
mv api/* api/batch
```

After moving the APIs to a new directory, the same needs to be applied to the controllers:

```bash
mkdir internal/controller/batch
mv internal/controller/* internal/controller/batch/
```

> 

Reviewer: remove the unnecessary spaces

Next, we'll need to update all the references to the old package name. 

> For CronJob, that'll be `main.go` and `controllers/batch/cronjob_controller.go`. 

Reviewer 2: ```suggestion
For CronJob, the paths to be replaced would be `main.go` and `controllers/batch/cronjob_controller.go` to their respective locations in the new project structure. 
```

If you've added additional files to your project, you'll need to track down
imports there as well.

Finally, fix the PROJECT file manually, the command `kubebuilder edit --multigroup=true`
sets our project to multigroup, but it doesn't fix the path of the existing APIs.

> For each resource need to modify the path and perform no additional modifications.

Reviewer 2: ```suggestion
For each resource we need to modify the path.
```

For instance, for a file:
                                                      
```yaml
# Code generated by tool. DO NOT EDIT.
# This file is used to track the info used to scaffold your project
# and allow the plugins properly work.
# More info: https://book.kubebuilder.io/reference/project-config.html
domain: tutorial.kubebuilder.io
layout:
- go.kubebuilder.io/v4
multigroup: true
projectName: test
repo: tutorial.kubebuilder.io/project
resources:
- api:
    crdVersion: v1
    namespaced: true
  controller: true
  domain: tutorial.kubebuilder.io
  group: batch
  kind: CronJob
  path: tutorial.kubebuilder.io/project/api/v1beta1
  version: v1beta1
version: "3"
```

Replace `path: tutorial.kubebuilder.io/project/api/v1beta1` for 
`path: tutorial.kubebuilder.io/project/api/batch/v1beta1`

> In this way, if the project is not new and has previous APIs already implemented will be in the previous structure. 
> Notice that with the `multi-group` project the Kind API's files are
> created under `api/<group>/<version>` instead of `api/<version>`. 
> Also, note that the controllers will be created under `internal/controller/<group>` instead of `internal/controller`. 
> That is the reason why we moved the previously generated APIs with the provided scripts in the previous steps. 
> Remember to update the references afterwards.

Reviewer 2: ```suggestion
In this process, if the project is not new and has previous implemented APIs they would still need to be modified as needed. 
Notice that with the `multi-group` project the Kind API's files are created under `api/<group>/<version>` instead of `api/<version>`. 
Also, note that the controllers will be created under `internal/controller/<group>` instead of `internal/controller`.
 
That is the reason why we moved the previously generated APIs to their respective locations in the new structure. 
Remember to update the references in imports accordingly.
```
